xwiki-platform-core/xwiki-platform-repository/xwiki-platform-repository-server-api/src/main/java/org/xwiki/repository/internal/RepositoryManager.java

```diff
@@ -1102,4 +1198,38 @@ private boolean updateCollection(BaseObject extensionObject, String fieldName, C
 
         return needSave;
     }
+
+    public BaseObject getExtensionVersionObject(XWikiDocument extensionDocument, String version) {
+        return getExtensionVersionObject(extensionDocument, version, true);
+    }
+
+    public BaseObject getExtensionVersionObject(XWikiDocument extensionDocument, String version, boolean allowProxying)
+    {
+        if (version == null) {
+            List<BaseObject> objects =
+                    extensionDocument.getXObjects(XWikiRepositoryModel.EXTENSIONVERSION_CLASSREFERENCE);
+
+            if (objects == null || objects.isEmpty()) {
+                return null;
+            } else {
+                return objects.get(objects.size() - 1);
+            }
+        }
+
+        BaseObject extensionVersionObject =
+                extensionDocument.getObject(XWikiRepositoryModel.EXTENSIONVERSION_CLASSNAME, "version", version, false);
+
+        if( extensionVersionObject == null && allowProxying && isVersionProxyingEnabled(extensionDocument) ){
+             //no ExtensionVersionClass object so we need to create such object temporarily
```

Comment:
No, again you should not save any document because it's super slow. I'm only talking about extensionDocument.clone() here so that you don't break the XWikiDocument instance coming from the cache since that's the issue here.